moveit_experimental/jog_arm/include/jog_arm/jog_ros_interface.h

```diff
@@ -0,0 +1,252 @@
+#include <sensor_msgs/JointState.h>
+#include <sensor_msgs/Joy.h>
+#include <std_msgs/Bool.h>
+#include <std_msgs/Float64MultiArray.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <tf2_ros/transform_listener.h>
+#include <thread>
+#include <trajectory_msgs/JointTrajectory.h>
+
+namespace jog_arm
+{
+// Variables to share between threads, and their mutexes
+struct JogArmShared
+{
+  pthread_mutex_t shared_variables_mutex;
+
+  geometry_msgs::TwistStamped command_deltas;
+
+  moveit_msgs::JogJoint joint_command_deltas;
+
+  sensor_msgs::JointState joints;
+
+  double collision_velocity_scale = 1;
+
+  // Indicates that an incoming Cartesian command is all zero velocities
+  bool zero_cartesian_cmd_flag = true;
+
+  // Indicates that an incoming joint angle command is all zero velocities
+  bool zero_joint_cmd_flag = true;
+
+  // Indicates that we have not received a new command in some time
+  bool command_is_stale = false;
+
+  // The new trajectory which is calculated
+  trajectory_msgs::JointTrajectory new_traj;
+
+  // Timestamp of incoming commands
+  ros::Time incoming_cmd_stamp = ros::Time(0.);
+
+  bool ok_to_publish = false;
+};
+
+// ROS params to be read.
+// See the yaml file in /config for a description of each.
+struct JogArmParameters
+{
+  std::string move_group_name, joint_topic, cartesian_command_in_topic, command_frame, command_out_topic,
+      planning_frame, warning_topic, joint_command_in_topic, command_in_type, command_out_type;
+  double linear_scale, rotational_scale, joint_scale, lower_singularity_threshold, hard_stop_singularity_threshold,
+      lower_collision_proximity_threshold, hard_stop_collision_proximity_threshold, low_pass_filter_coeff,
+      publish_period, publish_delay, incoming_command_timeout, joint_limit_margin, collision_check_rate;
+  bool use_gazebo, check_collisions, publish_joint_positions, publish_joint_velocities, publish_joint_accelerations;
+};
+
+/**
+ * Class JogROSInterface - Instantiated in main(). Handles ROS subs & pubs and
+ * creates the worker threads.
+ */
+class JogROSInterface
+{
+public:
+  JogROSInterface();
+
+private:
+  // ROS subscriber callbacks
+  void deltaCartesianCmdCB(const geometry_msgs::TwistStampedConstPtr& msg);
+  void deltaJointCmdCB(const moveit_msgs::JogJointConstPtr& msg);
+  void jointsCB(const sensor_msgs::JointStateConstPtr& msg);
+
+  bool readParameters(ros::NodeHandle& n);
+
+  // Jogging calculation thread
+  static bool startJogCalcThread(const JogArmParameters& parameters, JogArmShared& shared_variables,
+                                 const robot_model_loader::RobotModelLoaderPtr model_loader_ptr);
+
+  // Collision checking thread
+  static bool startCollisionCheckThread(const JogArmParameters& parameters, JogArmShared& shared_variables,
+                                        const robot_model_loader::RobotModelLoaderPtr& model_loader_ptr);
+
+  // Variables to share between threads
+  struct JogArmShared shared_variables_;
```

Comment:
``struct`` should not be here  (that's C-style i think)

Reply:
It's just a `class` but with all members public by default.

Reply:
I know what a struct is! :-)

Here he's doing a C-style declaration of a variable that is only for backwards compatibility 
https://stackoverflow.com/questions/7729646/using-struct-keyword-in-variable-declaration-in-c